opentreemap/treemap/instance.py

```diff
@@ -768,9 +771,39 @@ def raise_errors(errors):
         if errors:
             raise_errors(errors)
 
-    def save(self, *args, **kwargs):
+    @transaction.atomic
+    def _bulk_assign_task_reference_numbers(self, tasks):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for task in tasks:
+            task.reference_number = instance.task_sequence_number
+            instance.task_sequence_number += 1
+        instance.save(update_fields=['task_sequence_number'])
+
+    @transaction.atomic
+    def _bulk_assign_work_order_reference_numbers(self, work_orders):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for work_order in work_orders:
+            work_order.reference_number = instance.work_order_sequence_number
+            instance.work_order_sequence_number += 1
+        instance.save(update_fields=['work_order_sequence_number'])
+
+    def save(self, tasks=None, work_orders=None, *args, **kwargs):
+        """
+        Keyword Arguments:
+
+        tasks - List of Task models
+        work_orders - List of WorkOrder models
+        """
         self.full_clean()
 
         self.url_name = self.url_name.lower()
 
+        if tasks:
```

Comment:
The instance `save` method does not feel like the right place to put this operation. Could you explain the design constraints that would push us toward this approach rather than something like:


- Add an `atomic` helper method to `Instance` to get the next `n` reference numbers and update the `task_sequence_number` field appropriately (`n` would default to 1).
- Decorate `Task.save_with_user` as `atomic` and have it use its `instance` reference to call the helper and fetch the next sequence number.
- Add an atomic `Task.bulk_save_with_user` class method for when you want to create a set of tasks in a loop.